src/librustc_resolve/lib.rs

```diff
@@ -2056,6 +2056,7 @@ impl<'a> Resolver<'a> {
                         this.with_current_self_type(self_type, |this| {
                             for impl_item in impl_items {
                                 this.check_proc_macro_attrs(&impl_item.attrs);
+                                this.visit_generics(&impl_item.generics);
```

Comment:
Does this change have any unintended sife-effects?

Reply:
@sunjay 
You have to visit generics inside of `with_type_parameter_rib` here and for trait items.
This was done previously for methods in this piece of code below:
```rust
this.with_type_parameter_rib(type_parameters, |this| {
    visit::walk_impl_item(this, impl_item); // We visited method signature, `FnKind::Method` and its generics here previously
});
```

Reply:
Yes, this change is not quite right. =) 

The way that resolve works, it has these "ribs" that are in scope, those are the sets of names that can be used. When you *visit* the generics, you are then resolving the names that appear within. So when you visit the generics here, we are visiting without any names in scope.

What we want to do then is to move up the code that is specific to methods below, and execute it for *all* impl-items:

```rust
                                         let type_parameters =
                                            HasTypeParameters(&impl_item.generics,
                                                             MethodRibKind(!sig.decl.has_self()));
                                         this.with_type_parameter_rib(type_parameters, |this| {...}
```

in that case, we can `visit_generics` here, and you can .. probably forget what I said about modifying  `FnKind::Method` above.

Reply:
Actually, I don't think we want to call `visit_generics` here. It will happen as part of the `walk_impl_item` that we see below.

Reply:
@nikomatsakis I don't think I can actually move the `HasTypeParameters` code above because it uses `sig` which is only in scope for methods. After this refactoring, I may have to come back and add `HasTypeParameters` for `TraitItemKind::Type` in this match statement.